### DISCUSSION TITLE
Multi-line comment textobject

### DISCUSSION THREAD

**Discussion by julien-h2** [nvim-mini/mini.nvim#1234] [Q&A] [answered]
Hello, I would like to create a comment textobject that I can use inside mini.ai
I want to map this object to `q` so that `saq` select around the comment block.

I know neovim added a [comment textobject here](https://github.com/neovim/neovim/blob/f01c764cc6f82399edfa0d47a7bafbf7c95e2747/runtime/lua/vim/_comment.lua) but I don't know how to make it work with mini.ai mappings to go to next object, go to previous object, etc.

Also, I can't manage to map it to `q` because even if I map `q` to `gc`, when I type `saq` it selects around quotes, not comment.

P.S Using treesitter here does not seem to work at it only select a single line of comment, not the whole block.

**Comment by echasnovski** [comment, marked as answer]
Unfortunately, it is not an easy thing to do. The reason is because in order to make this work with "next"/"prev" search methods, 'mini.ai' needs to know about them when deciding which one to choose. The built-in commenting was designed to exclusively used interactively, so it doesn't provide any (stable) Lua API.

The best approach to solve what you are asking for is to manually write a textobject specification as a [function that returns an array of regions](https://github.com/echasnovski/mini.nvim/blob/e50cf9de614500a20e47cfc50e30a100042f91c3/doc/mini-ai.txt#L294-L313) representing all consecutive comment blocks. Depending on how robust and capable you want it to be, implementation can span from maybe about 15 lines to at least 50.

However, the most reasonable approach to solve the problem I'd say is to utilize [`'mini.bracketed' `comment` target](https://github.com/echasnovski/mini.nvim/blob/e50cf9de614500a20e47cfc50e30a100042f91c3/doc/mini-bracketed.txt#L38) to move across comment blocks and using available "current comment block" `gc` textobject.

**Comment by julien-h2** [comment]
I see. It would also be good to get the select around, select around next, etc. So mini.ai is preferable.

I've looked into the codebase and I'm thinking that I could make a function to merge the regions from `gen_spec.treesitter({ a = "@comment.outer", i = "@comment.inner" })` if they are consecutive. 

In mini.ai, is there a way to denote a visual-block selection? For instance everything after column 10 on lines 3, 4, 5? To match these kind of comments as a visual block:

```go
a = 5  // comment 1
b = 6  // comment 2
c = 7  // comment 3
```

**Reply by echasnovski** [reply]
> In mini.ai, is there a way to denote a visual-block selection?

Each region can have [`vis_mode` field](https://github.com/echasnovski/mini.nvim/blob/e50cf9de614500a20e47cfc50e30a100042f91c3/doc/mini-ai.txt#L192-L194) to denote that.

**Reply by julien-h2** [reply]
Thanks!
